File: lint-arwaky/src-python/contract/mcp_server_port.py
"""mcp_server_port — Port interface for MCP server lifecycle and tool management."""

from abc import ABC, abstractmethod
from typing import Any, Callable, Coroutine

from ..taxonomy import SymbolName, DescriptionVO


class IMcpServerPort(ABC):
    """Port for managing an MCP server instance."""

    @abstractmethod
    def register_tool(
        self,
        name: SymbolName,
        description: DescriptionVO,
        handler: Callable[..., Coroutine[Any, Any, Any]],
    ) -> None:
        """Register a tool with the MCP server."""
        ...

    @abstractmethod
    async def run_server(self) -> None:
        """Start the MCP server."""
        ...

    @abstractmethod
    def stop_server(self) -> None:
        """Stop the MCP server."""
        ...

File: lint-arwaky/src-rust/contract/mcp_server_port.rs
/// mcp_server_port — Port interface for MCP server lifecycle and tool management.
use crate::taxonomy::{DescriptionVO, SymbolName};
use std::future::Future;
use std::pin::Pin;

pub type ToolHandler = Box<
    dyn Fn(
            serde_json::Value,
        ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, String>> + Send>>
        + Send
        + Sync,
>;

#[async_trait::async_trait]
pub trait IMcpServerPort: Send + Sync {
    /// Register a tool with the MCP server.
    fn register_tool(
        &self,
        name: SymbolName,
        description: DescriptionVO,
        handler: ToolHandler,
    );

    /// Start the MCP server.
    async fn run_server(&self) -> Result<(), String>;

    /// Stop the MCP server.
    fn stop_server(&self);
}
